Support tensor.backward in dynamo

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172126
* __->__ #169415

Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan committed:
Support tensor.backward in dynamo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1008,14 +1008,80 @@ def wrap(i, sub_proxy):
-            hints=[*graph_break_hints.FUNDAMENTAL],
+    def method_backward(
+        self, gradient=None, retain_graph=None, create_graph=None, inputs=None
+    ):
```
Almost definitely want to config guard this for now

ezyang reviewed (changes requested):
You need to make sure that the autograd graph the backward() call calls doesn't extrude "beyond" the compiled region. Concretely, construct a test where a non-leaf requires grad True input is passed in and then try to call backward().

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1008,14 +1009,89 @@ def wrap(i, sub_proxy):
+                gb_type="Unsupported Tensor.backward() call",
+                context=f"call_method {self} backward {gradient} {retain_graph} {create_graph} {inputs}",
+                explanation="Dynamo currently does not support tracing `Tensor.backward()`.",
+                hints=[*graph_break_hints.FUNDAMENTAL],
```
Remove `FUNDAMENTAL` - replace with suggestion to turn on your config?

anijain2305 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1008,14 +1009,89 @@ def wrap(i, sub_proxy):
+            grad_args.append(gradient)
+
+        autograd_grad_fn = VariableTracker.build(tx, torch.autograd.grad)
+        grads_var = autograd_grad_fn.call_function(tx, grad_args, grad_kwargs)
```
> Because we make sure the compiled region is guarded enough that it only takes leaf tensors.

Where do we ensure that?

tugsbayasgalan replied:
I do it in the PR below it https://github.com/pytorch/pytorch/pull/169493 since backward impl depends on supporting autograd.grad

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan commented on benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_training.csv:
```diff
@@ -34,35 +34,35 @@ DistillGPT2,pass,7
 
 
-ElectraForCausalLM,pass,4
+ElectraForCausalLM,pass,3
```
Turned off by default

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several of small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1072,18 +1074,158 @@ def wrap(i: Any, sub_proxy: torch.fx.Proxy) -> VariableTracker:
+        Only collects leaf tensors (no grad_fn). Non-leaf tensors are skipped
+        (or error if error_on_non_leaf=True) since the accumulate_grad polyfill
+        cannot handle .grad access on non-leaf tensors as Dynamo creates generic
+        GetAttrVariable (https://github.com/pytorch/pytorch/issues/171204).
```
The real reason to not include non-leafs, is when you are automatically collecting inputs, you MUST NOT stop the gradients at non-leafs

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1072,18 +1074,158 @@ def wrap(i: Any, sub_proxy: torch.fx.Proxy) -> VariableTracker:
+                "tensor does not require grad and does not have a grad_fn"
+            )
+
+        # Step 1: Collect leaf tensors to compute gradients for
```
What would be safer here is to verify that you actually have a input source corresponding to every AccumulateGrad grad fn that you flood fill from the outputs passed to grad.

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1072,18 +1074,158 @@ def wrap(i: Any, sub_proxy: torch.fx.Proxy) -> VariableTracker:
+            # or parameters that are created in forward.
+            all_vars = chain(
+                tx.output.leaf_var_creation_order,
+                tx.output.input_source_to_var.values(),
```
This works, but only because torch.autograd.grad handling is very conservative. And it turns out it MUST be conservative, because see the convo at https://github.com/pytorch/pytorch/pull/169493/files#r2665611139 But it would be better NOT to ever generate illegal torch.autograd.grad call (where there are not enough grad inputs) in the first place, by doing extra checks here.

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan commented:
@pytorchbot merge -i

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor.backward in dynamo"


Following https://github.com/pytorch/pytorch/issues/167729, one natural extension is that we should able to trace through tensor.backward. When we do tensor.backward, we rewrite it as autograd.grad + torch.ops.inductor.accumulate_grad_ in dynamo. When there is no explicit parameters being passed in to backward, we just look at dynamo graph inputs and find all input parameters. This works because autograd.grad will just return None for the unused ones. 

The change looks quite big but there are several small/independent fixes:
1. Have to put allow_unused parameter to autograd.grad when we are automatically query-ing parameters because we are passing in all possible parameters that may or may not correspond to particular autograd.grad call 
2. We try to collect in-graph created parameters 
3. We use node proxies to deduplicate parameters that might be aliasing each other
4. If user passes in non-leaf parameter to backward, we graph break because generic .grad handling in dynamo is limited today. Will do as a follow up. 
2. If we are calling backward on a tensor with no grad fn, we should error. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
